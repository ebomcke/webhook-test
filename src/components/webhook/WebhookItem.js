import React, { Component } from 'react';
import { Feed, Icon, Segment, Modal } from 'semantic-ui-react';
import { compose } from 'recompose';
import { withFirestore } from 'react-firestore';
import { withNavigation } from '../../contexts/Navigation';
import Timestamp from '../base/Timestamp';
import { withAccount } from '../../contexts/Account';

class WebhookItem extends Component {
  constructor(props) {
    super(props);
    this.deleteWebhook = this.deleteWebhook.bind(this);
  }

  async deleteWebhook() {
    const { webhook, firestore } = this.props;
    await firestore
      .collection('webhooks')
      .doc(webhook.id)
      .delete();
  }

  shortBody() {
    const { webhook } = this.props;
    let body = JSON.stringify(JSON.parse(webhook.body), null, 2);
    let lines = body.split(/\n/);
    if (lines.length > 5) {
      return lines
        .slice(0, 5)
        .concat('[...]', lines.slice(-1))
        .join('\r\n');
    }
    return body;
  }

  render() {
    const { webhook, account } = this.props;
    return (
      <Feed.Event>
        <Feed.Label>
          <Icon name="code" />
        </Feed.Label>
        <Feed.Content>
          <Feed.Summary>
            {`/${account.organisationName}/${webhook.path}`}
            <Feed.Date>
              <Timestamp timestamp={webhook.date} prefix="Received" />
            </Feed.Date>
          </Feed.Summary>
          <Feed.Extra text>
            <Modal
              size="large"
              trigger={
                <Segment inverted>
                  <pre>
                    <code>{this.shortBody()}</code>
                  </pre>
                </Segment>
              }
            >
              <Modal.Header>Full message</Modal.Header>
              <Modal.Content scrolling>
                <Modal.Description>
                  <Segment inverted>
                    <pre>
                      <code>
                        {JSON.stringify(JSON.parse(webhook.body), null, 2)}
                      </code>
                    </pre>
                  </Segment>
                </Modal.Description>
              </Modal.Content>
            </Modal>
          </Feed.Extra>
          <Feed.Meta>
            <a name='delete-webhook-link' onClick={this.deleteWebhook}>
              <Icon name="remove" />
              Remove
            </a>
          </Feed.Meta>
        </Feed.Content>
      </Feed.Event>
    );
  }
}

export default compose(
  withNavigation,
  withFirestore,
  withAccount,
)(WebhookItem);
