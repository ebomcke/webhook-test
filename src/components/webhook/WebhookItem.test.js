import React from 'react';
import { create } from 'react-test-renderer';
import WebhookItem from './WebhookItem';
jest.mock('../base/Timestamp');

import mockWebhooks from '../../contexts/Webhooks';

describe('Rendering tests', () => {
  test('Initial rendering - full body', () => {
    const webhookItem = create(<WebhookItem webhook={mockWebhooks[0]} />);
    expect(webhookItem.toJSON()).toMatchSnapshot();
  });

  test('Initial rendering - shortened body', () => {
    const webhookItem = create(<WebhookItem webhook={mockWebhooks[1]} />);
    expect(webhookItem.toJSON()).toMatchSnapshot();
  });
});
