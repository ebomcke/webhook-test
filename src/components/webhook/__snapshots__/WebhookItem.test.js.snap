// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests Initial rendering - full body 1`] = `
<div
  className="event"
>
  <div
    className="label"
  >
    <i
      aria-hidden="true"
      className="code icon"
    />
  </div>
  <div
    className="content"
  >
    <div
      className="summary"
    >
      /unit-test-org/webhook-path-1
      <div
        className="date"
      >
        <div>
          This is a placeholder for the Timestamp.
          <pre>
            <code>
              {"timestamp":"webhook-date-1","prefix":"Received"}
            </code>
          </pre>
        </div>
      </div>
    </div>
    <div
      className="text extra"
    >
      <div
        className="ui inverted segment"
      >
        <pre>
          <code>
            {
  "unitTest": true
}
          </code>
        </pre>
      </div>
    </div>
    <div
      className="meta"
    >
      <a
        name="delete-webhook-link"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="remove icon"
        />
        Remove
      </a>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Initial rendering - shortened body 1`] = `
<div
  className="event"
>
  <div
    className="label"
  >
    <i
      aria-hidden="true"
      className="code icon"
    />
  </div>
  <div
    className="content"
  >
    <div
      className="summary"
    >
      /unit-test-org/webhook-path-2
      <div
        className="date"
      >
        <div>
          This is a placeholder for the Timestamp.
          <pre>
            <code>
              {"timestamp":"webhook-date-2","prefix":"Received"}
            </code>
          </pre>
        </div>
      </div>
    </div>
    <div
      className="text extra"
    >
      <div
        className="ui inverted segment"
      >
        <pre>
          <code>
            {
  "unitTest": true,
  "first": "line",
  "second": "line",
  "third": "line",
[...]
}
          </code>
        </pre>
      </div>
    </div>
    <div
      className="meta"
    >
      <a
        name="delete-webhook-link"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="remove icon"
        />
        Remove
      </a>
    </div>
  </div>
</div>
`;
